esphome/components/ee895/ee895.cpp

```diff
@@ -0,0 +1,139 @@
+#include "ee895.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace ee895 {
+
+static const char *const TAG = "ee895";
+
+static const uint16_t CRC16_ONEWIRE_START = 0xFFFF;
+static const uint8_t FUNCTION_CODE_READ = 0x03;
+static const uint16_t SERIAL_NUMBER = 0x0000;
+static const uint16_t TEMPERATURE_ADDRESS = 0x03EA;
+static const uint16_t CO2_ADDRESS = 0x0424;
+static const uint16_t PRESSURE_ADDRESS = 0x04B0;
+
+void EE895Component::setup() {
+  uint16_t crc16_check = 0;
+  ESP_LOGCONFIG(TAG, "Setting up EE895...");
+  write_command_(SERIAL_NUMBER, 8);
+  uint8_t serial_number[20];
+  this->read(serial_number, 20);
+
+  crc16_check = (serial_number[19] << 8) + serial_number[18];
+  if (crc16_check != calc_crc16_(serial_number, 19)) {
+    this->error_code_ = CRC_CHECK_FAILED;
+    this->mark_failed();
+    return;
+  }
+  uint32_t serial_number1 =
+      (serial_number[2] << 24) + (serial_number[3] << 16) + (serial_number[4] << 8) + serial_number[5];
+  uint32_t serial_number2 =
+      (serial_number[6] << 24) + (serial_number[7] << 16) + (serial_number[8] << 8) + serial_number[9];
+  uint32_t serial_number3 =
+      (serial_number[10] << 24) + (serial_number[11] << 16) + (serial_number[12] << 8) + serial_number[13];
+  uint32_t serial_number4 =
+      (serial_number[14] << 24) + (serial_number[15] << 16) + (serial_number[16] << 8) + serial_number[17];
+  ESP_LOGV(TAG, "    Serial Number: 0x%08X%08X%08X%08X", serial_number1, serial_number2, serial_number3,
+           serial_number4);
```

Comment:
I just found this out, but there's a `format_hex()` function in the core.
```suggestion
  ESP_LOGV(TAG, "    Serial Number: 0x%s", format_hex(serial_number + 2, 16).c_str());
```